0002-ether-htlc-dynamic-final-addresses.md

```diff
@@ -0,0 +1,121 @@
+
+## Research
+
+Bitcoin requires you witness the HTLC output with the secret and a signature (or a just the refund signature after a timeout).
+The transaction that has the witness will specify where to redeem the funds in its output.
+
+To follow the same pattern in Ethereum, our redeem transaction will need in its data:
+
+1. The secret
+2. The address to give the money to (the final redeem address)
+3. a signature under the redeem identity from the COMIT node
+
+The refund would be the same without (1)
+
+### The Final Address
+
+The decision we have to make here is whether the final address is:
+
+1. Provided in the data
+2. The sender address
+
+(2) is nice because it means that there is no way to mess it up and send a transaction to an address you don't own. It also saves a bit of gas.
+
+On the other hand, that might be a feature.
+You may want to send the asset in the HTLC to a contract or something, so I'm leaning towards (1).
+
+
+### The signature
+
+The key things when deciding what to sign are:
+
+1. It must not be possible to front run the transaction, learn the secret, and then create another transaction that redeems to a different address.
+2. It must not be possible to re-use the signature on a different chain or a different contract on the same chain.
+
+To ensure (1) we need to sign the final address (redeem or refund) with the private key of the corresponding identity.
+To ensure (2) we should sign the HTLC's contract address, the network id and chain id.
+
+So I propose that we sign the following messages for each outcome:
+
+- Redeem: `network_id || chain_id || htlc_contract_address || final_redeem_address`
+- Refund: `network_id || chain_id || htlc_contract_address || final_refund_address`
+
+We have to hash the message before signing it.
+In this case we can choose any 256 bit output hash function because we control both the singing code (in COMIT) and the verifier (in the contract).
+Our choices are `SHA-256` and `Keccak-256`.
+I think we should just use `Keccak-256` because it's less gas on EVM.
+
+### Verifying the signature
+
+To verify the signature we use the `ecrecover` precompiled contract.
+Note that this means the signature we need to send is the 65 byte recoverable ECDSA signature.
+(A recoverable signature is one where you can figure out the public key from the signature itself)
+
+Here's the psuedocode for the contract:
+
+``` rust
+let (expiry, redeem_identity, refund_identity, secret_hash) = embedded_somehow...;
+
+// We're redeemimg with secret
+if CALLDATASIZE == (20 + 65 + 32) {
+    let (final_redeem_address, signature, secret) = CALLDATA();
+
+    if sha256(secret) == secret_hash {
+        let message_hash = sha3(network_id, chain_id, address(), final_redeem_address);
+        let (v,r,s) = signature;
+        if redeem_identity == ecrecover(message_hash, v, r, s) {
+            log("Redeemed(secret)"); // could also log the address it was redeemed to
```

Comment:
`log("Redeemed(secret)");` is correct as we need to print the secret to extract it easily from the transaction receipt.

Reply:
Yeah I wasn't sure if we cared about the address it was redeemed to in terms of transparency. With Bitcoin we can obviously tell where the redeem went to, if we want to be able to do the same here we need to log the address in the data too.
I'm in favour of logging the address as well.